UIKit/S2MRefreshControl.m

```diff
@@ -0,0 +1,220 @@
+    }
+
+}
+
+- (void)animateWithFractionDragged:(CGFloat)fractionDragged
+{
+    if ([self.loadingView conformsToProtocol:@protocol(S2MControlLoadingView)]) {
+        id<S2MControlLoadingView> controlLoadingView = (id<S2MControlLoadingView>)self.loadingView;
+        [controlLoadingView animateWithFractionDragged:fractionDragged];
+    }else if([self.loadingView isKindOfClass:[UIImageView class]]){
+        self.loadingView.alpha = 1;
+        self.loadingView.transform = CGAffineTransformMakeRotation(2*M_PI * MAX(0.0, fractionDragged));
+    }else{
+        NSAssert(false, @"Expected loadingView to conforms to S2MControlLoadingView protocol or subclass UIImageView");
+    }
+}
+
+- (void)beginRefreshing
+{
+    if (self.isRefreshing) {
+        return;
+    }
+    
+    self.scrollViewInitialInsets = self.scrollView.contentInset;
+    self.isRefreshing = YES;
+    
+    self.scrollView.contentInset = UIEdgeInsetsMake(self.startLoadingThreshold, 0, 0, 0);
+    self.scrollView.scrollEnabled = NO;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self sendActionsForControlEvents:UIControlEventValueChanged];
+        [self startAnimating];
+        self.scrollView.scrollEnabled = YES;
+    });
+}
+
+- (void)endRefreshing
+{
+    if (!self.isRefreshing) {
+        [self stopAnimating];
+        return;
+    }
+    
+    [UIView animateWithDuration:0.2 delay:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
+        self.loadingView.alpha = 0;
+    } completion:^(BOOL finished) {
+        [UIView animateWithDuration:0.3 delay:0.1 options:UIViewAnimationOptionAllowUserInteraction animations:^{
+            [self.scrollView setContentOffset:CGPointZero];
+            self.scrollView.contentInset = self.scrollViewInitialInsets;
+        } completion:^(BOOL finished2) {
+            [self stopAnimating];
+            self.isRefreshing = NO;
+        }];
+    }];
+}
+
+#pragma mark - ScrollView
+
+- (void)containingScrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
+{
+    // do nothing
+}
+
+- (void)containingScrollViewDidScroll:(UIScrollView *)scrollView
+{
+    CGFloat offset = scrollView.contentOffset.y + scrollView.contentInset.top;
+    if (offset <= 0.0 && !self.isRefreshing && !scrollView.isDecelerating) {
+        CGFloat fractionDragged = -offset/self.startLoadingThreshold;
+        [self animateWithFractionDragged:fractionDragged];
+        if (fractionDragged >= 1.0) {
+            [self beginRefreshing];
+        }
+    }
+}
+
+#pragma mark - KVO
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+    if ([keyPath isEqual:NSStringFromSelector(@selector(delegate))]) {
+        id newDelegate = change[NSKeyValueChangeNewKey];
+        self.originalScrollViewDelegate = newDelegate;
+    }
+}
+
+
+#pragma mark - UIScrollViewDelegate
+//// We forward delegate calls
+
+- (BOOL)respondsToSelector:(SEL)aSelector
```

Comment:
yes it is. see https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtForwarding.html